man/cat1/arraystat.1



ARRAYSTAT(1)             TIPSY COMMANDS              ARRAYSTAT(1)



NAME
     arraystat - print statistics about the array values

SYNOPSIS
     arraystat _b_o_x-_n_u_m_b_e_r _p_a_r_t_i_c_l_e

DESCRIPTION
     arraystat calculates the mean, median, and upper quartile of
     the  loaded  array values for the particles of type _p_a_r_t_i_c_l_e
     in box _b_o_x-_n_u_m_b_e_r. An array is loaded  using  the  readarray
     command.  Boxes can be created and loaded using the makebox,
     setbox, or setsphere commands.  If no array is loaded or  if
     the box _b_o_x-_n_u_m_b_e_r is not loaded an error is generated.

OPTIONS
  Particle
     The possible choices for _p_a_r_t_i_c_l_e are:

          all
             Use the array values of all  the  particles  in  box
             _b_o_x-_n_u_m_b_e_r when calculating the statistics.

          baryon
             Use the array values of only the star and gas parti-
             cles  in box _b_o_x-_n_u_m_b_e_r when calculating the statis-
             tics.

          dark
             Use the array values of only the dark matter  parti-
             cles  in box _b_o_x-_n_u_m_b_e_r when calculating the statis-
             tics.

          gas
             Use the array values of only the  gas  particles  in
             box _b_o_x-_n_u_m_b_e_r when calculating the statistics.

          star
             Use the array values of only the star  particles  in
             box _b_o_x-_n_u_m_b_e_r when calculating the statistics.

SEE ALSO
     boxstat(1),  makebox(1),   readarray(1),   setbox(1),   set-
     sphere(1), viewarray(1)












KQ Release 2.0     Last change: 22 MARCH 1994                   1





code/ChangeLog
Wed Jan 11 13:09:36 1995  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (SRCS): added alloca.c so it will be included in the
	distribution.

Wed Dec 28 13:42:47 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* readmark.c: close file after it is read.

	* main.c: RedrawViewport(): make can_size the maximum of
	can_height and can_width.

	* canvas_resize_proc.c: Same as above.

Mon Dec  5 09:39:44 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* xray_lum_load.c, star_lum.c, star_lum_redshift.c: Changed
	argument to my_gets() to blank string.

Fri Dec  2 08:21:50 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (READLINE_SRC): delete config.h.
	(tipsy): added dependence on ALLOCA.

	* fits.c: Added test for endianness.

Thu Dec  1 15:14:58 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* vista.c: fixed initialization of quantity[].

	* find_shape.c: Forced inertia_cm to be symmetric.  This avoid
	problems with jacobi().

Tue Nov  1 11:26:56 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in : fits_xray.c is now gone.  fits() is used instead.
	defs.h, xray.c, vista.c, fits.c and arguments.c are also changed.

Mon Oct 31 09:43:29 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* main.c: RedrawColorbar(): fixed off-by-one drawing errors.

Wed Oct 12 14:27:16 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Created version 2.1.2.

	* absorb.c: New function for calculating aborbtion spectra.

Tue Sep  6 16:06:02 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline/: Replaced with readline-2.0 from prep.ai.mit.edu.

	* my_gets.c: Changed cast to match new external from readline/.

Fri Jul 29 10:25:49 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Ensure that the axes projections do not overflow.

Fri Jul 15 17:28:15 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* profile.c: Check to see if fopen() fails.

Wed Jul 13 09:46:18 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h: Decreased INTMAX to 2^29-1.  (One bit for sign, one bit
	to do INTMAX - (-INTMAX), and one bit for zoom out.

Tue Jul 12 15:55:17 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* canvas_resize_proc.c: Call plot_sub() and draw_label() for
	servers that automatically do backing store and don't generate an
	expose event.

	* main.c: ZoomIn(), ZoomOut(), Pan(), SetBox(): Ignore button
	presses in inactive windows.
	
	* main.c, window.c: Changed ResizeRequest to ConfigureNotify.

Sat Jun  4 13:12:06 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* fits.c, and fits_xray.c: Changed write() to fwrite() for
	portability.

Thu May  5 10:53:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Increase clipping range of velocities to -2*INTMAX <
	v < 2*INTMAX.
	* defs.h: Change INTMAX to 2^30-1 to accommodate the above
	clipping.

Tue Apr 19 18:00:14 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* comm.h: NSK: added pcenter and writeboxrot commands.

Mon Apr 18 14:14:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Add "unzoom" button.

Fri Apr 15 16:01:56 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Fixed bug with color of marked particles.
	Deleted extra conversion to screen coordinates.
	
	* maxpos.c: Make the margins a little smaller.

Thu Mar 31 14:03:11 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Changed XDrawArc to XFillArc for vector draws.

Fri Mar  4 18:09:25 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Added Neal's changes for read_vector() and "master".

Mon Feb 28 11:51:39 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* openascii.c: Added code from Peter Teuben to uncompress files.

Thu Feb 24 13:21:50 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Changed current_plot logical to current_project and
	current_color.

Tue Feb 22 11:24:05 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline: Updated to readline from gdb-1.12

Thu Feb 17 11:46:07 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Implimented a software "z-buffer".  Only draw
	particles with the highest particle color.

Wed Feb 16 13:21:35 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Merged plot_all and plot_vel into one routine.  A
	new routine all_color() is created.  Changes were also needed in
	the view_*() routines.

Sat Nov 13 11:35:01 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Fixed typo for increasing ndark.

Wed Nov  3 14:35:56 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Deleted skip of XClearWindow on first time through.

Thu Oct 28 10:23:50 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* hard.c: Added movie parameter.

Tue Oct 26 11:29:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* alloca.c: New file in case system doesn't provide alloca().

	* vista.c: Included <stdlib.h> for malloc() definition.

Mon Oct 25 19:16:52 1993  Tom Quinn  (trq@virgo)

	* main.c: Attempt to find a depth 8 PseudoColor visual.

Thu Oct 21 08:56:32 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Set backgrounds of gc_color[] to black.

Tue Oct 19 10:56:36 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h, main.h, make_macro.c, read_macro.c, write_macro.c
	run_macro.c: re-vamped macro storage using linked lists.

Mon Oct 18 15:11:34 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* draworient.c: Added cast from Arg to long.

Fri Oct 15 10:04:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Added a check for index < 1.

	* plot_sub.c: Added a check for binary_loaded.
	Added check for boxes_loaded[0].

Wed Oct  6 09:13:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* zoom_window.c: Added zoom command.  Also changed comm.h and
	arguments.c.

Tue Oct  5 17:27:43 1993  Tom Quinn  (trq@hermes.astro.washington.edu

code/arraystat.c
/* $Header$
 * $Log$
 * Revision 1.2  1995/03/24 18:49:02  trq
 * Added "revrain" colormap.
 *
 * divv.c: included malloc.h.
 *
 * Revision 1.1.1.1  1995/01/10  22:57:36  trq
 * Import to CVS
 *
 * Revision 1.3  94/04/20  08:44:54  trq
 * Added title variable.
 * 
 * Revision 1.1  94/02/16  13:51:25  trq
 * Initial revision
 * 
 */
#include "defs.h"
#include <malloc.h>

void input_error();
static Real *array_select();

void
arraystat(job)
    char *job;
{
    char command[MAXCOMM] ;
    int box;
    char type[MAXCOMM] ;
    double mean = 0.0;
    double median;
    double uquartile;
    int i, offset;
    int totpart;
    int aoffset;
    Real **medtmp;

    if(sscanf(job,"%s %d %s",command,&box,type) == 3) {
	if (boxes_loaded[box])
	  {
	    if(array_size != 0)
	      {
		if (strcmp(type,"gas") == 0 ) {
		  medtmp = (Real **)malloc(boxlist[box].ngas*sizeof(*medtmp));
		  for (i = 0 ;i < boxlist[box].ngas ;i++)
		    {
		      aoffset = boxlist[box].gp[i]-gas_particles;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  mean /= boxlist[box].ngas;
		  uquartile = *(array_select(medtmp, boxlist[box].ngas,
				    3*boxlist[box].ngas/4));
		  median = *(array_select(medtmp, boxlist[box].ngas,
				    boxlist[box].ngas/2));
		  free(medtmp);
		  printf("<array mean, median and upper quartile of the gas in box %d>\n", box);
		}
		else if (strcmp(type,"dark") == 0 ) {
		  offset = header.nsph;
		  medtmp = (Real **)malloc(boxlist[box].ndark*sizeof(*medtmp));
		  for (i = 0 ;i < boxlist[box].ndark ;i++)
		    {
		      aoffset = boxlist[box].dp[i]-dark_particles+offset;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  mean /= boxlist[box].ndark;
		  uquartile = *(array_select(medtmp, boxlist[box].ndark,
				    3*boxlist[box].ndark/4));
		  median = *(array_select(medtmp, boxlist[box].ndark,
				    boxlist[box].ndark/2));
		  free(medtmp);
		  printf("<array mean, median and upper quartile of the dark matter in box %d>\n",
			 box);
		}
		else if (strcmp(type,"star") == 0 ) {
		  offset = header.nsph + header.ndark;
		  medtmp = (Real **)malloc(boxlist[box].nstar*sizeof(*medtmp));
		  for (i = 0 ;i < boxlist[box].nstar ;i++)
		    {
		      aoffset = boxlist[box].sp[i]-star_particles+offset;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  mean /= boxlist[box].nstar;
		  uquartile = *(array_select(medtmp, boxlist[box].nstar,
				    3*boxlist[box].nstar/4));
		  median = *(array_select(medtmp, boxlist[box].nstar,
				    boxlist[box].nstar/2));
		  free(medtmp);
		  printf("<array mean, median and upper quartile of the stars in box %d>\n",
			 box);
		}
		else if (strcmp(type,"baryon") == 0 ) {
		  medtmp = (Real **)malloc((boxlist[box].ngas+boxlist[box].nstar)
				  *sizeof(*medtmp));
		  for (i = 0 ;i < boxlist[box].ngas ;i++)
		    {
		      aoffset = boxlist[box].gp[i]-gas_particles;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  offset = header.nsph + header.ndark;
		  for (i = 0 ;i < boxlist[box].nstar ;i++)
		    {
		      aoffset = boxlist[box].sp[i]-star_particles+offset;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  totpart = boxlist[box].nstar + boxlist[box].ngas;
		  mean /= totpart;
		  uquartile = *(array_select(medtmp, totpart, 3*totpart/4));
		  median = *(array_select(medtmp, totpart, totpart/2));
		  free(medtmp);
		  printf("<array mean, median and upper quartile of the baryons in box %d>\n",
			 box);
		}
		else if (strcmp(type,"all") == 0 ) {
		  medtmp = (Real **)malloc((boxlist[box].ngas+boxlist[box].nstar
				   +boxlist[box].ndark)*sizeof(*medtmp));
		  for (i = 0 ;i < boxlist[box].ngas ;i++)
		    {
		      aoffset = boxlist[box].gp[i]-gas_particles;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  offset = header.nsph;
		  for (i = 0 ;i < boxlist[box].ndark ;i++)
		    {
		      aoffset = boxlist[box].dp[i]-dark_particles+offset;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  offset = header.nsph + header.ndark;
		  for (i = 0 ;i < boxlist[box].nstar ;i++)
		    {
		      aoffset = boxlist[box].sp[i]-star_particles+offset;
		      medtmp[i] = &array[aoffset];
		      mean += array[aoffset];
		    }
		  totpart = boxlist[box].nstar + boxlist[box].ngas
		    + boxlist[box].ndark;
		  mean /= totpart;
		  uquartile = *(array_select(medtmp, totpart, 3*totpart/4));
		  median = *(array_select(medtmp, totpart, totpart/2));
		  free(medtmp);
		  printf("<array mean, median and upper quartile of box %d>\n",
			 box);
		}
		else {
		    printf("<sorry, %s is not a proper data type, %s>\n",type,title) ;
		}
		printf("%g %g %g\n", mean, median, uquartile);
	      }
	    else
	      {
		printf("<sorry, no array loaded, %s>\n",title);
	      }
	}
	else {
	    printf("<sorry, box %d is not loaded, %s>\n",box,title) ;
	}
	  
    }
    else {
	input_error(command) ;
    }
}

static
void swap(p, q)
     Real **p;
     Real **q;
{
  Real *tmp = *p;
  *p = *q;
  *q = tmp;
}

static Real *array_select(a, n, k)
     Real **a;
     int n, k;
{
  int l = 0;
  int u = n-1;
  while(l < u)
    {
      int m;
      int i;
      
      swap(&a[l], &a[(l+u)/2]);
      m = l;

      for(i = l+1; i <= u; i++)
	{
	  if(*a[i] < *a[l])
	    {
	      m++;
	      swap(&a[m], &a[i]);
	    }
	}
      swap(&a[l], &a[m]);
      if(k <= m) u = m;
      if(k >= m) l = m+1;
    }
  return a[k];
}
